### PR TITLE
Support defining the type when extending a factory with params()

### PR DESCRIPTION
**Pull request by lo1tuma**
Fixes: #71

### CODE REVIEW

File: lib/factory.ts [hunk 1]
```diff
@@ -124,10 +124,12 @@ export class Factory<T, I = any> {
...
    * @returns a new factory
    */
-  params(params: DeepPartial<T>): this {
+  params<TOverride extends T = T>(
+    params: DeepPartial<TOverride>,
+  ): Factory<TOverride, I> {
     const factory = this.clone();
     factory._params = merge({}, this._params, params, mergeCustomizer);
-    return factory;
+    return factory as unknown as Factory<TOverride, I>;
```
Comment (stevehanson) [inline lib/factory.ts:132]: The issue, and I'm not sure yet how to solve it, is that even though `TOverride` extends `T`, we aren't guaranteeing that new params are passed in such that the built object is actually a `TOverride`.

File: lib/__tests__/factory-builders.test.ts [hunk 2]
```diff
@@ -226,6 +231,38 @@ describe('params', () => {
     expect(adminFactory.build().admin).toBe(true);
     expect(adminFactory.build().admin).toBe(true);
   });
+
+  it('adds parameters for a sub-type which are then also accepted in build()', () => {
+    const adminFactory = userFactory.params<AdminUser>({
+      admin: true,
```
Comment (stevehanson) [inline lib/__tests__/factory-builders.test.ts:237]: My concern is that if `admin: true` were not supplied, then the factory happily compiles but does not actually return an `AdminUser`.

**Review by stevehanson — CHANGES_REQUESTED** [review]
Thanks for your contribution. There are a couple issues here that circumvent the type-checker which would prevent me from accepting this as is. We will need a solution that does not do a type coercion (eg. `as unknown as TOverride`) so that we can guarantee that types are working correctly.

File: lib/factory.ts (hunk 1 above)
Comment (lo1tuma) [reply lib/factory.ts:132]: You are absolutely right, `params` is `Partial` and might not contain all properties required for `TOverride`. 🤔 I also can’t think about any other approach to make it work with the `params()` method. Maybe with a dedicated `extends()` method that accepts another generator function.

**Comment by lo1tuma** [comment]
Thanks for your feedback. I’m closing the PR as the idea doesn’t work as I thought.

**Pull request closed without merging**